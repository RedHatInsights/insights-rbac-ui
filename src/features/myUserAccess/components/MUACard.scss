.rbac-c-mua-bundles__card--applist {
  text-transform: capitalize;
}
.rbac-c-mua-bundles__cardlink {
  color: inherit;
  &:hover {
    text-decoration: none;
  }
  &--disabled {
    pointer-events: none;
  }
}
