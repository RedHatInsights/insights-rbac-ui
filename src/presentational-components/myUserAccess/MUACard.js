import React from 'react';
import { NavLink } from 'react-router-dom';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { routes } from '../../../package.json';

import { Stack, StackItem, Title, Card, CardTitle, CardBody } from '@patternfly/react-core';

import { bundleData } from './bundles';
import useSearchParams from '../../hooks/useSearchParams';

const MUACard = ({ header, entitlements, isDisabled }) => {
  const { bundle: bundleParam } = useSearchParams('bundle');
  return (
    <React.Fragment>
      <Title headingLevel="h3" size="xl">
        {header}
      </Title>
      <Stack hasGutter>
        {entitlements &&
          entitlements?.map(([key]) => {
            const data = bundleData.find(({ entitlement }) => entitlement === key);
            return data ? (
              <StackItem key={key}>
                <NavLink className="ins-c-mua-bundles__cardlink" to={{ pathname: routes['my-user-access'], search: `bundle=${key}` }}>
                  <Card
                    key={data.title}
                    isFlat={isDisabled || key !== bundleParam}
                    isSelectable={!isDisabled}
                    isSelected={!isDisabled && key === bundleParam}
                    className={classNames('ins-c-mua-bundles__card', `ins-c-mua-bundles__card--${data.entitlement}`, {
                      'ins-c-mua-bundles__card--disabled': isDisabled,
                    })}
                  >
                    <CardTitle className="ins-c-mua-bundles__card--header"> {data.title}</CardTitle>
                    <CardBody>
                      <Stack hasGutter>
                        <StackItem className="ins-c-mua-bundles__card--applist">
                          {Object.entries(data.apps || {}).map(([appName]) => (
                            <span key={appName}> {appName} </span>
                          ))}
                        </StackItem>
                      </Stack>
                    </CardBody>
                  </Card>
                </NavLink>
              </StackItem>
            ) : null;
          })}
      </Stack>
    </React.Fragment>
  );
};

MUACard.propTypes = {
  header: PropTypes.string,
  entitlements: PropTypes.array,
  isDisabled: PropTypes.bool,
};

export default MUACard;
