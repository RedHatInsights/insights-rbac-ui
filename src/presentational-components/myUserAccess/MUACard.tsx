import React, { useState } from 'react';
import { NavLink } from 'react-router-dom';
import classNames from 'classnames';
import { Card, CardBody, CardHeader, CardTitle, List, ListItem, Stack, StackItem, Title } from '@patternfly/react-core';
import { useFlag } from '@unleash/proxy-client-react';

import { bundleData } from './bundles';
import useSearchParams from '../../hooks/useSearchParams';

import './MUACard.scss';

interface MUACardProps {
  header?: string;
  entitlements?: Array<[string, any]>;
  isDisabled?: boolean;
}

interface BundleData {
  entitlement: string;
  title: string;
  apps: Record<string, string | undefined>;
  appsIds: string[];
}

const MUACard: React.FC<MUACardProps> = ({ header, entitlements = [], isDisabled = false }) => {
  const { bundle: bundleParam } = useSearchParams('bundle') as { bundle?: string };
  const [, setIsChecked] = useState<string>('');

  const isITLess = useFlag('platform.rbac.itless');

  const onChange = (event: React.FormEvent<HTMLInputElement>) => {
    setIsChecked((event.target as HTMLInputElement).id);
  };

  const bundles: BundleData[] = isITLess ? bundleData.filter((data) => data.entitlement === 'rhel') : bundleData;

  return (
    <React.Fragment>
      {header && (
        <Title headingLevel="h3" size="xl">
          {header}
        </Title>
      )}
      <Stack
        className={classNames({
          'pf-v5-u-mt-lg': !header,
        })}
        hasGutter
      >
        {entitlements &&
          bundles?.map((data) => {
            const entitlementEntry = entitlements.find(([key]) => data.entitlement === key);
            const isEntitled = entitlementEntry && entitlementEntry[1]?.is_entitled;
            const key = data.entitlement;
            return isEntitled ? (
              <StackItem key={key} className="rbac-c-mua-cardWrapper">
                <NavLink
                  aria-label="card-link"
                  className={classNames('rbac-c-mua-bundles__cardlink', { 'pf-v5-u-background-color-disabled-color-300': isDisabled })}
                  to={{ pathname: '', search: `bundle=${key}` }}
                >
                  <Card
                    ouiaId={`${data.title}-card`}
                    key={data.title}
                    isFlat={isDisabled || key !== bundleParam}
                    isSelectable={!isDisabled}
                    isSelected={!isDisabled && key === bundleParam}
                    className={classNames({
                      'pf-v5-u-background-color-disabled-color-300': isDisabled,
                    })}
                  >
                    <CardHeader
                      selectableActions={{
                        selectableActionId: '',
                        name: data.title,
                        variant: 'single',
                        onChange,
                      }}
                    >
                      <CardTitle className="pf-v5-u-font-weight-light" data-ouia-component-id={`${data.title}-card-title`}>
                        {data.title}
                      </CardTitle>
                    </CardHeader>

                    <CardBody data-ouia-component-id={`${data.title}-card-body`}>
                      <List className="pf-v5-u-color-400 pf-v5-u-font-size-sm rbac-c-mua-bundles__card--applist" isPlain>
                        {Object.entries(data.apps || {}).map(([appName]) => (
                          <ListItem key={appName}> {appName} </ListItem>
                        ))}
                      </List>
                    </CardBody>
                  </Card>
                </NavLink>
              </StackItem>
            ) : null;
          })}
      </Stack>
    </React.Fragment>
  );
};

export default MUACard;
