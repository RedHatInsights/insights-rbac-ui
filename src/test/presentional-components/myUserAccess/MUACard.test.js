import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import MUACard from '../../../presentational-components/myUserAccess/MUACard';
import { MemoryRouter } from 'react-router-dom';
import { Card } from '@patternfly/react-core';
/**
 * Mock bundle data
 */
jest.mock('../../../presentational-components/myUserAccess/bundles');

const RouterWrapper = ({ children }) => <MemoryRouter>{children}</MemoryRouter>;
const entitlementsMock = [
  ['entitled', { is_entitled: true, is_trial: false }],
  ['not-entitled', { is_entitled: false, is_trial: false }],
];
describe('<MUACard />', () => {
  it('should render correctly', () => {
    expect(
      toJson(
        mount(
          <RouterWrapper>
            <MUACard header="Test" entitlements={entitlementsMock} />
          </RouterWrapper>
        ).find(MUACard)
      )
    ).toMatchSnapshot();
  });

  it('should render correctly disabled', () => {
    expect(
      toJson(
        mount(
          <RouterWrapper>
            <MUACard header="Test" isDisabled entitlements={entitlementsMock} />
          </RouterWrapper>
        ).find(MUACard)
      )
    ).toMatchSnapshot();
  });

  it('should not render a MUACard if not listed in entitlemends', () => {
    const wrapper = mount(
      <RouterWrapper>
        <MUACard header="Test" isDisabled entitlements={[]} />
      </RouterWrapper>
    );
    expect(wrapper.find(Card)).toHaveLength(0);
  });
});
