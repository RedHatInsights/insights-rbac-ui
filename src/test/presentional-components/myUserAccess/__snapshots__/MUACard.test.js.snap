// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MUACard /> should render correctly 1`] = `
<MUACard
  entitlements={
    Array [
      Array [
        "entitled",
        Object {
          "is_entitled": true,
          "is_trial": false,
        },
      ],
      Array [
        "not-entitled",
        Object {
          "is_entitled": false,
          "is_trial": false,
        },
      ],
    ]
  }
  header="Test"
>
  <Title
    headingLevel="h3"
    size="xl"
  >
    <h3
      className="pf-c-title pf-m-xl"
    >
      Test
    </h3>
  </Title>
  <Stack
    hasGutter={true}
  >
    <div
      className="pf-l-stack pf-m-gutter"
    >
      <StackItem
        className="ins-c-mua-cardWrapper"
        key="entitled"
      >
        <div
          className="pf-l-stack__item ins-c-mua-cardWrapper"
        >
          <NavLink
            className="ins-c-mua-bundles__cardlink"
            to={
              Object {
                "pathname": "/my-user-access",
                "search": "bundle=entitled",
              }
            }
          >
            <Link
              aria-current={null}
              className="ins-c-mua-bundles__cardlink"
              to={
                Object {
                  "pathname": "/my-user-access",
                  "search": "bundle=entitled",
                }
              }
            >
              <LinkAnchor
                aria-current={null}
                className="ins-c-mua-bundles__cardlink"
                href="/my-user-access?bundle=entitled"
                navigate={[Function]}
              >
                <a
                  aria-current={null}
                  className="ins-c-mua-bundles__cardlink"
                  href="/my-user-access?bundle=entitled"
                  onClick={[Function]}
                >
                  <Card
                    className="ins-c-mua-bundles__card ins-c-mua-bundles__card--entitled"
                    isFlat={true}
                    isSelectable={true}
                    isSelected={false}
                    key="Entitled"
                  >
                    <article
                      className="pf-c-card pf-m-selectable pf-m-flat ins-c-mua-bundles__card ins-c-mua-bundles__card--entitled"
                      data-ouia-component-id={0}
                      data-ouia-component-type="PF4/Card"
                      data-ouia-safe={true}
                      tabIndex="0"
                    >
                      <CardTitle
                        className="ins-c-mua-bundles__card--header"
                      >
                        <div
                          className="pf-c-card__title ins-c-mua-bundles__card--header"
                        >
                           
                          Entitled
                        </div>
                      </CardTitle>
                      <CardBody>
                        <div
                          className="pf-c-card__body"
                        >
                          <Stack
                            hasGutter={true}
                          >
                            <div
                              className="pf-l-stack pf-m-gutter"
                            >
                              <StackItem
                                className="ins-c-mua-bundles__card--applist"
                              >
                                <div
                                  className="pf-l-stack__item ins-c-mua-bundles__card--applist"
                                >
                                  <span
                                    key="eApp1"
                                  >
                                     
                                    eApp1
                                     
                                  </span>
                                  <span
                                    key="eApp2"
                                  >
                                     
                                    eApp2
                                     
                                  </span>
                                </div>
                              </StackItem>
                            </div>
                          </Stack>
                        </div>
                      </CardBody>
                    </article>
                  </Card>
                </a>
              </LinkAnchor>
            </Link>
          </NavLink>
        </div>
      </StackItem>
      <StackItem
        className="ins-c-mua-cardWrapper"
        key="not-entitled"
      >
        <div
          className="pf-l-stack__item ins-c-mua-cardWrapper"
        >
          <NavLink
            className="ins-c-mua-bundles__cardlink"
            to={
              Object {
                "pathname": "/my-user-access",
                "search": "bundle=not-entitled",
              }
            }
          >
            <Link
              aria-current={null}
              className="ins-c-mua-bundles__cardlink"
              to={
                Object {
                  "pathname": "/my-user-access",
                  "search": "bundle=not-entitled",
                }
              }
            >
              <LinkAnchor
                aria-current={null}
                className="ins-c-mua-bundles__cardlink"
                href="/my-user-access?bundle=not-entitled"
                navigate={[Function]}
              >
                <a
                  aria-current={null}
                  className="ins-c-mua-bundles__cardlink"
                  href="/my-user-access?bundle=not-entitled"
                  onClick={[Function]}
                >
                  <Card
                    className="ins-c-mua-bundles__card ins-c-mua-bundles__card--not-entitled"
                    isFlat={true}
                    isSelectable={true}
                    isSelected={false}
                    key="Not entitled"
                  >
                    <article
                      className="pf-c-card pf-m-selectable pf-m-flat ins-c-mua-bundles__card ins-c-mua-bundles__card--not-entitled"
                      data-ouia-component-id={1}
                      data-ouia-component-type="PF4/Card"
                      data-ouia-safe={true}
                      tabIndex="0"
                    >
                      <CardTitle
                        className="ins-c-mua-bundles__card--header"
                      >
                        <div
                          className="pf-c-card__title ins-c-mua-bundles__card--header"
                        >
                           
                          Not entitled
                        </div>
                      </CardTitle>
                      <CardBody>
                        <div
                          className="pf-c-card__body"
                        >
                          <Stack
                            hasGutter={true}
                          >
                            <div
                              className="pf-l-stack pf-m-gutter"
                            >
                              <StackItem
                                className="ins-c-mua-bundles__card--applist"
                              >
                                <div
                                  className="pf-l-stack__item ins-c-mua-bundles__card--applist"
                                >
                                  <span
                                    key="neApp1"
                                  >
                                     
                                    neApp1
                                     
                                  </span>
                                  <span
                                    key="neApp2"
                                  >
                                     
                                    neApp2
                                     
                                  </span>
                                </div>
                              </StackItem>
                            </div>
                          </Stack>
                        </div>
                      </CardBody>
                    </article>
                  </Card>
                </a>
              </LinkAnchor>
            </Link>
          </NavLink>
        </div>
      </StackItem>
    </div>
  </Stack>
</MUACard>
`;

exports[`<MUACard /> should render correctly disabled 1`] = `
<MUACard
  entitlements={
    Array [
      Array [
        "entitled",
        Object {
          "is_entitled": true,
          "is_trial": false,
        },
      ],
      Array [
        "not-entitled",
        Object {
          "is_entitled": false,
          "is_trial": false,
        },
      ],
    ]
  }
  header="Test"
  isDisabled={true}
>
  <Title
    headingLevel="h3"
    size="xl"
  >
    <h3
      className="pf-c-title pf-m-xl"
    >
      Test
    </h3>
  </Title>
  <Stack
    hasGutter={true}
  >
    <div
      className="pf-l-stack pf-m-gutter"
    >
      <StackItem
        className="ins-c-mua-cardWrapper"
        key="entitled"
      >
        <div
          className="pf-l-stack__item ins-c-mua-cardWrapper"
        >
          <NavLink
            className="ins-c-mua-bundles__cardlink ins-c-mua-bundles__cardlink--disabled"
            to={
              Object {
                "pathname": "/my-user-access",
                "search": "bundle=entitled",
              }
            }
          >
            <Link
              aria-current={null}
              className="ins-c-mua-bundles__cardlink ins-c-mua-bundles__cardlink--disabled"
              to={
                Object {
                  "pathname": "/my-user-access",
                  "search": "bundle=entitled",
                }
              }
            >
              <LinkAnchor
                aria-current={null}
                className="ins-c-mua-bundles__cardlink ins-c-mua-bundles__cardlink--disabled"
                href="/my-user-access?bundle=entitled"
                navigate={[Function]}
              >
                <a
                  aria-current={null}
                  className="ins-c-mua-bundles__cardlink ins-c-mua-bundles__cardlink--disabled"
                  href="/my-user-access?bundle=entitled"
                  onClick={[Function]}
                >
                  <Card
                    className="ins-c-mua-bundles__card ins-c-mua-bundles__card--entitled ins-c-mua-bundles__card--disabled"
                    isFlat={true}
                    isSelectable={false}
                    isSelected={false}
                    key="Entitled"
                  >
                    <article
                      className="pf-c-card pf-m-flat ins-c-mua-bundles__card ins-c-mua-bundles__card--entitled ins-c-mua-bundles__card--disabled"
                      data-ouia-component-id={2}
                      data-ouia-component-type="PF4/Card"
                      data-ouia-safe={true}
                    >
                      <CardTitle
                        className="ins-c-mua-bundles__card--header"
                      >
                        <div
                          className="pf-c-card__title ins-c-mua-bundles__card--header"
                        >
                           
                          Entitled
                        </div>
                      </CardTitle>
                      <CardBody>
                        <div
                          className="pf-c-card__body"
                        >
                          <Stack
                            hasGutter={true}
                          >
                            <div
                              className="pf-l-stack pf-m-gutter"
                            >
                              <StackItem
                                className="ins-c-mua-bundles__card--applist"
                              >
                                <div
                                  className="pf-l-stack__item ins-c-mua-bundles__card--applist"
                                >
                                  <span
                                    key="eApp1"
                                  >
                                     
                                    eApp1
                                     
                                  </span>
                                  <span
                                    key="eApp2"
                                  >
                                     
                                    eApp2
                                     
                                  </span>
                                </div>
                              </StackItem>
                            </div>
                          </Stack>
                        </div>
                      </CardBody>
                    </article>
                  </Card>
                </a>
              </LinkAnchor>
            </Link>
          </NavLink>
        </div>
      </StackItem>
      <StackItem
        className="ins-c-mua-cardWrapper"
        key="not-entitled"
      >
        <div
          className="pf-l-stack__item ins-c-mua-cardWrapper"
        >
          <NavLink
            className="ins-c-mua-bundles__cardlink ins-c-mua-bundles__cardlink--disabled"
            to={
              Object {
                "pathname": "/my-user-access",
                "search": "bundle=not-entitled",
              }
            }
          >
            <Link
              aria-current={null}
              className="ins-c-mua-bundles__cardlink ins-c-mua-bundles__cardlink--disabled"
              to={
                Object {
                  "pathname": "/my-user-access",
                  "search": "bundle=not-entitled",
                }
              }
            >
              <LinkAnchor
                aria-current={null}
                className="ins-c-mua-bundles__cardlink ins-c-mua-bundles__cardlink--disabled"
                href="/my-user-access?bundle=not-entitled"
                navigate={[Function]}
              >
                <a
                  aria-current={null}
                  className="ins-c-mua-bundles__cardlink ins-c-mua-bundles__cardlink--disabled"
                  href="/my-user-access?bundle=not-entitled"
                  onClick={[Function]}
                >
                  <Card
                    className="ins-c-mua-bundles__card ins-c-mua-bundles__card--not-entitled ins-c-mua-bundles__card--disabled"
                    isFlat={true}
                    isSelectable={false}
                    isSelected={false}
                    key="Not entitled"
                  >
                    <article
                      className="pf-c-card pf-m-flat ins-c-mua-bundles__card ins-c-mua-bundles__card--not-entitled ins-c-mua-bundles__card--disabled"
                      data-ouia-component-id={3}
                      data-ouia-component-type="PF4/Card"
                      data-ouia-safe={true}
                    >
                      <CardTitle
                        className="ins-c-mua-bundles__card--header"
                      >
                        <div
                          className="pf-c-card__title ins-c-mua-bundles__card--header"
                        >
                           
                          Not entitled
                        </div>
                      </CardTitle>
                      <CardBody>
                        <div
                          className="pf-c-card__body"
                        >
                          <Stack
                            hasGutter={true}
                          >
                            <div
                              className="pf-l-stack pf-m-gutter"
                            >
                              <StackItem
                                className="ins-c-mua-bundles__card--applist"
                              >
                                <div
                                  className="pf-l-stack__item ins-c-mua-bundles__card--applist"
                                >
                                  <span
                                    key="neApp1"
                                  >
                                     
                                    neApp1
                                     
                                  </span>
                                  <span
                                    key="neApp2"
                                  >
                                     
                                    neApp2
                                     
                                  </span>
                                </div>
                              </StackItem>
                            </div>
                          </Stack>
                        </div>
                      </CardBody>
                    </article>
                  </Card>
                </a>
              </LinkAnchor>
            </Link>
          </NavLink>
        </div>
      </StackItem>
    </div>
  </Stack>
</MUACard>
`;
